### DISCUSSION TITLE
Integrate GLSP server with Sirius editor

### DISCUSSION THREAD

**Discussion by Avaritia1** [eclipse-glsp/glsp#966] [Q&A]
I am new to GLSP and trying to integrate a custom-built Sirius graphical editor to GLSP server. I already found examples for GLSP connected to Theia and Eclipse with EMF but I cannot yet make a connection between Sirius editor and GLSP server. Can someone help with this? Any pointers are appreciated.

**Comment by planger** [comment]
Hi,

an integration of GLSP and Sirius can be done on different levels, depending on how much you'd like to reuse from your existing Sirius-based implementation. Of course this is a trade-off between reuse and complexity of the final solution.
The different levels are:

1. You can integrate the GLSP server with GMF runtime: thus a GLSP server would actually load the Sirius model similar to how it is loaded in Sirius right now and translate GMF's notation model instance into a GModel to send it down to the client for rendering. On changes triggered in the client, the GLSP server would apply GMF commands to the loaded model and then update the client. This allows to reuse also much of the existing implementation's diagram behavior. But it adds quite a lot of complexity in the implementation and a lot of additional resource cost during runtime. We actually did something similar a few years ago as a proof of concept for GMF runtime (it was Papyrus though and not Sirius).

https://user-images.githubusercontent.com/588090/227186351-355ba8d1-4daa-4627-b44b-0c4f7c2a664b.mp4


2. You can integrate the GLSP server with EMF (model level, not GMF on diagram level). With that approach the GLSP server would simply load the Sirius models as EMF models and use them as [GLSP source models](https://www.eclipse.org/glsp/documentation/sourcemodel/). This much less complexity and there are reusable components to do that and it is pretty much standard GLSP development without much overhead. See also the [GLSP project template for EMF](https://github.com/eclipse-glsp/glsp-examples/tree/master/project-templates/java-emf-theia).

3. If it is a use case, you could write a generator from an odesign file to generate a GLSP server and client. Conceptually this would be possible, but I never really looked into doing that in detail.

For the sake of having a maintainable solution for the years to come, I tend towards approach 2, but in the end it depends a lot on your existing Sirius editor and how much you accept complexity at runtime and dependency in order to achieve a high reuse or whether you'd be willing to re-implement diagram behavior in GLSP natively to avoid extra dependencies and resource cost at runtime.

Let me know what you think and feel free to share more details, so that we can guide you more concretely in the best direction.

**Reply by Avaritia1** [reply]
Thank you for the reply! Yes, I also think approach 2 is better for me. I have already built the [java-emf-eclipse example](https://github.com/eclipse-glsp/glsp-examples/tree/master/project-templates/java-emf-eclipse) on my laptop. Do you have any suggestions for next steps? Thanks!

**Reply by planger** [reply]
I would suggest loading your EMF models (probably *.aird) instead of the *.tasklist model in [your source model](https://github.com/eclipse-glsp/glsp-examples/blob/master/project-templates/java-emf-eclipse/glsp-server/org.eclipse.glsp.example.javaemf.server/src/org/eclipse/glsp/example/javaemf/server/model/TaskListSourceModelStorage.java) and adjust [your GModel factory](https://github.com/eclipse-glsp/glsp-examples/blob/master/project-templates/java-emf-eclipse/glsp-server/org.eclipse.glsp.example.javaemf.server/src/org/eclipse/glsp/example/javaemf/server/model/TaskListGModelFactory.java) to produce a diagram structure ([GModel](https://www.eclipse.org/glsp/documentation/gmodel/)) to [be rendered](https://www.eclipse.org/glsp/documentation/rendering/) from it.

Once that is up and running for your diagram, I suggest to start working on the [operation handlers](https://github.com/eclipse-glsp/glsp-examples/tree/master/project-templates/java-emf-eclipse/glsp-server/org.eclipse.glsp.example.javaemf.server/src/org/eclipse/glsp/example/javaemf/server/handler) to allow creating, editing, and deleting model elements.

**Reply by Avaritia1** [reply]
Hi Langer! I already created a Sirius Graphical Editor with creating, editing and deleting model elements functions. The functions are in the Palette(As shown below). Now I can run this graphical editor on my laptop. And I want to integrate it to GLSP server. Is it doable? Could you please share some suggestions? Thanks!

![image](https://user-images.githubusercontent.com/80332537/230490809-fc790db2-b7e7-45e3-a572-660a67e71dae.png)

**Reply by planger** [reply]
Hi @Avaritia1,

What do you mean by "integrate it to GLSP server"? Do you mean you would want to have a GLSP-based editor similar to the Sirius-based editor you have right now? Is it required that it works with the same model resources (*.statemachine)?
Or do you want a GLSP editor to be integrated in your existing editor alongside the Sirius diagram editor?

As mentioned above there is no off-the-shelf example implementation for integrating Sirius and GLSP at the moment. So I can't point to any existing code but can just give general suggestions of how this could be done and what the initial steps are. But I mentioned them already above.

Overall, as far as I understand your use case, this is certainly possible. Depending on the level of integration (as mentioned above), the EMF-integration with GLSP may be sufficient if you want to develop a GLSP editor for *.statemachine models. If your Sirius editor and GLSP editor shall coexist in the future, you'd need to also support reading the aird files to keep the locations, sizes, etc. of model elements in the diagram.

But to give more concrete suggestions, I'd need more information of what you eventually want to achieve.

**Reply by Avaritia1** [reply]
Thank for the reply! Yes, I wonder if this is possible:  Client -- Sirius graphical editor(which I have created) and make it connect with GLSP server. If this approach is not very feasible or takes a lot of time. Then I would consider using standard GLSP development, ie -- building a GLSP-based editor similar to the Sirius-based editor I have right now.

And yes, it will work with the same model sources (*.statemachine).

**Reply by planger** [reply]
Ah ok, thanks for the clarification!

Well the combination connecting Sirius as a client to a GLSP server is rather difficult to achieve, especially because the entire user interaction isn't really built for the interaction patterns of the GLSP protocol and is pretty tied to the GMF runtime. It sure is possible but I'm not sure it is a pragmatic direction and only worth going if you have already a very complex and customized diagram client implementation that you'd need to preserve at all cost.

I think writing a dedicated the client and connecting the GLSP server to your domain model (*.statemachine) seems much more straightforward to me.

I hope that helps!

**Reply by Avaritia1** [reply]
Thanks for the suggestion! I will consider your suggestion on my project.
